minimint-api/src/db/mod.rs

```diff
@@ -124,8 +158,129 @@ impl<'a> dyn Database + 'a {
         })
     }
 
-    pub fn apply_batch(&self, batch: DbBatch) -> Result<()> {
-        self.raw_apply_batch(batch)
+    pub fn transaction<'a>(&'a self) -> Result<Box<dyn Transaction<'a>>> {
+        self.transaction()
+    }
+}
+
+impl<'a> dyn Transaction<'a> {
+    pub fn insert_entry<K>(&mut self, key: &K, value: &K::Value) -> Result<Option<K::Value>>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        match self.raw_insert_entry(&key.to_bytes(), value.to_bytes())? {
+            Some(old_val_bytes) => {
+                trace!(
+                    "insert_entry: Decoding {} from bytes {:?}",
+                    std::any::type_name::<K::Value>(),
+                    old_val_bytes
+                );
+                Ok(Some(K::Value::from_bytes(&old_val_bytes)?))
+            }
+            None => Ok(None),
+        }
+    }
+
+    pub fn insert_new_entry<K>(&mut self, key: &K, value: &K::Value) -> Result<()>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        self.raw_insert_new_entry(&key.to_bytes(), value.to_bytes())?;
+
+        trace!(
+            "insert_new_entry: Decoding {} from bytes {:?}",
+            std::any::type_name::<K::Value>(),
+            old_val_bytes
+        );
+        Ok(())
+    }
+
+    pub fn get_value<K>(&self, key: &K) -> Result<Option<K::Value>>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        let key_bytes = key.to_bytes();
+        let value_bytes = match self.raw_get_value(&key_bytes)? {
+            Some(value) => value,
+            None => return Ok(None),
+        };
+
+        trace!(
+            "get_value: Decoding {} from bytes {:?}",
+            std::any::type_name::<K::Value>(),
+            value_bytes
+        );
+        Ok(Some(K::Value::from_bytes(&value_bytes)?))
+    }
+
+    pub fn remove_entry<K>(&mut self, key: &K) -> Result<()>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        self.raw_remove_entry(&key.to_bytes())?;
+
+        trace!(
+            "remove_entry: Decoding {} from bytes {:?}",
+            std::any::type_name::<K::Value>(),
+            value_bytes
+        );
+        Ok(())
+    }
+
+    pub fn maybe_remove_entry<K>(&mut self, key: &K) -> Result<Option<K::Value>>
```

Comment:
Is it necessary to have a `maybe` / normal distinction? If the caller does not require the value to be there they are free to ignore the `Option`, if they do - to add `expect` or `ok_or_else`.

Reply:
Yes you are correct and initially I agreed. The database operations are supposed to know to an extent what is being overridden and replaced and ultimately having the maybe makes enabling the API a lot easier because that's how it works with batchItem and it can allow us to abstract some of the error checking from other areas of the code making other places cleaner and more readable.